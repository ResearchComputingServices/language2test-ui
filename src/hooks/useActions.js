import { useDispatch } from 'react-redux';
import _ from 'lodash';
import { actions as storeActions } from '../redux/slices';

export default function useActions(storeName) {
    const actions = storeActions[storeName];
    const dispatch = useDispatch();
    return _.reduce(actions, (accumulator, functionDefinition, functionName) => {
        accumulator[functionName] = (...args) => dispatch(functionDefinition(...args));
        return accumulator;
    }, {});
}
