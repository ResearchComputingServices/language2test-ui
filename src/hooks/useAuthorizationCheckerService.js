import useProvider from './useProvider';

export default function useAuthorizationCheckerService() {
    return useProvider('authorizationChecker')();
}
