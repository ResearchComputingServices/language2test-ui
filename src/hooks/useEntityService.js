import useService from './useService';

// TODO why do we have so many definitions in so many different places
export default function useEntityService(entity) {
    const entityServiceMap = {
        user: {
            service: useService('user'),
            identifierKey: 'name',
        },
        student: {
            service: useService('student'),
            identifierKey: 'name',
        },
        test: {
            service: useService('test'),
            identifierKey: 'name',
        },
        roles: {
            service: useService('roles'),
            identifierKey: 'name',
        },
        testCategory: {
            service: useService('testCategory'),
            identifierKey: 'name',
        },
        testType: {
            service: useService('testType'),
            identifierKey: 'name',
        },
        userFieldType: {
            service: useService('userFieldType'),
            identifierKey: 'name',
        },
        demographicQuestionnaireField: {
            service: useService('demographicQuestionnaireField'),
            identifierKey: 'display',
        },
        studentClass: {
            service: useService('studentClass'),
            identifierKey: 'display',
        },
        enumeration: {
            service: useService('enumeration'),
            identifierKey: 'name',
        },
    };
    return entityServiceMap[entity];
}
