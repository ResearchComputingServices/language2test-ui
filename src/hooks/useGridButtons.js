import _ from 'lodash';
import useRolesCheckerService from './useRolesCheckerService';

function getButton(rolesCheckerService, title, handler, options = {}) {
    const skipAuthorization = _.get(options, 'skipAuthorization', false);
    const authorization = _.get(options, 'authorization', null);
    const hasAuthorization = skipAuthorization ? true : (authorization ? (rolesCheckerService.has(authorization)) : true);
    if (_.isBoolean(handler) && !handler) {
        return {
            title,
            handler,
            ...options,
            disabled: true,
        };
    }
    if (!_.isFunction(handler) || !hasAuthorization) {
        return null;
    }
    return {
        title,
        handler,
        ...options,
    };
}

export default function useGridButtons(
    id, handlers, authorizations = {}, skipAuthorization = false,
) {
    const rolesCheckerService = useRolesCheckerService();
    const { create } = handlers;
    return [
        _.isNil(id) ? getButton(rolesCheckerService, 'Create', create, {
            authorization: authorizations.create,
            skipAuthorization,
        }) : null,
    ];
}
