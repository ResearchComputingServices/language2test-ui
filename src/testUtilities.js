/* eslint-disable import/prefer-default-export */
/* eslint-disable prefer-const */
/* eslint-disable react/prop-types */
/* eslint-disable react/no-children-prop */
import React from 'react';
import { shallow, mount as mountComponent } from 'enzyme';
import { Router } from 'react-router-dom';
import toJSON from 'enzyme-to-json';
import { Provider } from 'react-redux';
import configureMockStore from 'redux-mock-store';
import _ from 'lodash';
import historyService from './services/HistoryService';
import { reducers } from './redux/slices';

const mockStore = configureMockStore({ reducers });

function WrapperComponent({ propsFunc, children, Component }) {
    const props = propsFunc();
    return (
        <Component
            children={children}
            {...props}
        />
    );
}

export function createTest(Component, opts = {}, callback) {
    let { props, children, mount, state } = opts;
    if (_.isNil(props)) props = {};
    if (_.isNil(children)) children = null;
    if (mount) {
        return () => {
            state = state || {};
            const mockMath = Object.create(global.Math);
            mockMath.random = () => 0.5;
            global.Math = mockMath;
            const mountedTree = mountComponent((
                <Router history={historyService.getHistory()}>
                    <Provider store={mockStore({ ...state })}>
                        <Component {...props} >
                            {children}
                        </Component>
                    </Provider>
                </Router>));
            _.isFunction(callback) && callback(mountedTree);
        };
    }
    if (_.isFunction(props)) {
        return () => {
            const shallowTree = shallow(
                <WrapperComponent
                    Component={Component}
                    propsFunc={props}
                >
                    {children}
                </WrapperComponent>,
            ).dive();
            _.isFunction(callback) && callback(shallowTree);
        };
    }
    return () => {
        const shallowTree = shallow((
            <Component {...props} >
                {children}
            </Component>
        ));
        _.isFunction(callback) && callback(shallowTree);
    };
}

export function createSnapshotTest(Component, opts = {}) {
    return createTest(Component, opts, tree => expect(toJSON(tree)).toMatchSnapshot());
}
