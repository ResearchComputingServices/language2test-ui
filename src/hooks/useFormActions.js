import _ from 'lodash';
import { ToastsStore } from 'react-toasts';
import FileSaver from 'file-saver';
import { useEffect, useRef, useCallback } from 'react';
import useService from './useService';
import useNavigateHistoryByPattern from './useNavigateHistoryByPattern';
import useStore from './useStore';
import useActions from './useActions';

export default function(entity, locale) {
    if (_.isNil(locale)) {
        locale = entity;
    }
    const dataRef = useRef(null);
    const service = useService(entity);
    const cancel = useNavigateHistoryByPattern();
    const { confirmed: dialogConfirmed } = useStore('dialog');
    const dialogActions = useActions('dialog');

    const removeEntity = useCallback(async data => {
        try {
            await service.remove(data);
            dialogActions.hideDialog();
            ToastsStore.success(`Successfully deleted ${locale}`);
            cancel();
        } catch (err) {
            dialogActions.hideDialog();
            ToastsStore.error(`Failed to delete ${locale}`);
        }
    }, [cancel, dialogActions, locale, service]);

    useEffect(() => {
        if (dialogConfirmed) {
            removeEntity(dataRef.current);
        }
    }, [dialogConfirmed, removeEntity]);

    const create = async data => {
        if (_.isEmpty(data)) return;
        dataRef.current = data;
        try {
            const result = await service.add(data);
            ToastsStore.success(`Successfully created ${locale}`);
            return result;
        } catch (err) {
            ToastsStore.error(`Failed to create ${locale}`);
        }
    };

    const update = async data => {
        if (_.isEmpty(data)) return;
        dataRef.current = data;
        try {
            const result = await service.update(data);
            ToastsStore.success(`Successfully updated ${locale}`);
            return result;
        } catch (err) {
            ToastsStore.error(`Failed to update ${locale}`);
        }
    };

    const remove = async data => {
        if (_.isEmpty(data)) return;
        dataRef.current = data;
        dialogActions.showDialog({ title: `Are you sure you want to delete this ${locale}?` });
    };

    const download = async (data, type, extension) => {
        try {
            const { id } = data;
            type = !_.isNil(type) ? type : 'application/vnd.openxmlformats-officedocument.spreadsheetml.sheet';
            extension = !_.isNil(extension) ? extension : 'xlsx';
            const blob = new Blob(
                [await service.export(id)],
                { type },
            );
            FileSaver.saveAs(blob, `${_.kebabCase(locale)}-${id}.${extension}`);
        } catch (err) {
            ToastsStore.error(`Failed to export ${locale}`);
        }
    };

    return {
        create,
        update,
        remove,
        download,
        cancel,
    };
}
