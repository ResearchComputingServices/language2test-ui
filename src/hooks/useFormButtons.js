import _ from 'lodash';
import useAuthorizationCheckerService from './useAuthorizationCheckerService';

function getButton(authorizationCheckerService, title, handler, options = {}) {
    const authorization = _.get(options, 'authorization', null);
    const hasAuthorization = (authorization ? (authorizationCheckerService.has(authorization)) : true);
    if (_.isBoolean(handler) && !handler) {
        return {
            title,
            handler,
            ...options,
            disabled: true,
        };
    }
    if (!_.isFunction(handler) || !hasAuthorization) {
        return null;
    }
    return {
        title,
        handler,
        ...options,
    };
}

export default function useFormButtons(id, entity, handlers, disableMutation = false) {
    const authorizationCheckerService = useAuthorizationCheckerService();
    const { create, update, remove, download, cancel } = handlers;
    entity = _.kebabCase(entity);
    return [
        _.isNil(id) ? getButton(authorizationCheckerService, 'Create', create, {
            disabled: disableMutation,
            authorization: `create-${entity}`,
        }) : null,
        _.isNil(id) ? null : getButton(authorizationCheckerService, 'Update', update, {
            disabled: disableMutation,
            authorization: `update-${entity}`,
        }),
        _.isNil(id) ? null : getButton(authorizationCheckerService, 'Delete', remove, {
            disabled: disableMutation,
            authorization: `delete-${entity}`,
        }),
        _.isNil(id) ? null : getButton(authorizationCheckerService, 'Export', download, { authorization: `export-${entity}` }),
        getButton(authorizationCheckerService, 'Cancel', cancel, { type: 'utility' }),
    ];
}
