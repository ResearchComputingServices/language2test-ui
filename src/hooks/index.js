import { useMount, useUnmount, useMountedState } from 'react-use';
import { useForm } from 'react-hook-form';
import {
    useEffect,
    useState,
    useRef,
    useCallback,
    useMemo,
} from 'react';
import useTimeCounter from './useTimeCounter';
import useActions from './useActions';
import useStore from './useStore';
import useStoreAttribute from './useStoreAttribute';
import useWindowSize from './useWindowSize';
import useTestWizardStores from './useTestWizardStores';
import useTestWizardActions from './useTestWizardActions';
import useList from './useList';
import useData from './useData';
import useService from './useService';
import useInJestChecker from './useInJestChecker';
import useNavigateHistoryByPattern from './useNavigateHistoryByPattern';
import useIsVisible from './useIsVisible';
import useTraceUpdate from './useTraceUpdate';
import useUnload from './useUnload';
import useGridColumns from './useGridColumns';
import useFormLayout from './useFormLayout';
import useFormData from './useFormData';
import useEntityService from './useEntityService';
import useEntities from './useEntities';
import useTypeahead from './useTypeahead';
import useFormActions from './useFormActions';
import useIsWideScreenMode from './useIsWideScreenMode';
import usePluralize from './usePluralize';
import useProvider from './useProvider';
import useRoutes from './useRoutes';
import useDrawer from './useDrawer';
import useFormButtons from './useFormButtons';
import useAuthorizationCheckerService from './useAuthorizationCheckerService';
import useEventListener from './useEventListener';
import useGridActions from './useGridActions';
import usePagination from './usePagination';
import useDialog from './useDialog';

export {
    useEffect,
    useState,
    useRef,
    useCallback,
    useMemo,
    useUnload,
    useTimeCounter,
    useActions,
    useStore,
    useStoreAttribute,
    useWindowSize,
    useTestWizardStores,
    useTestWizardActions,
    useList,
    useData,
    useService,
    useProvider,
    useForm,
    useInJestChecker,
    useMount,
    useUnmount,
    useMountedState,
    useNavigateHistoryByPattern,
    useIsVisible,
    useTraceUpdate,
    useGridColumns,
    useFormLayout,
    useFormData,
    useEntityService,
    useEntities,
    useTypeahead,
    useFormActions,
    useIsWideScreenMode,
    usePluralize,
    useRoutes,
    useDrawer,
    useFormButtons,
    useAuthorizationCheckerService,
    useEventListener,
    useGridActions,
    usePagination,
    useDialog,
};
