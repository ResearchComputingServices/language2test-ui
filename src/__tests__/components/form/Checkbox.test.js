import { useForm } from '../../../hooks';
import Component from '../../../components/form/fields/Checkbox';
import { createSnapshotTest } from '../../../testUtilities';

describe('Checkbox', () => {
    it('should match snapshot', createSnapshotTest(Component, {
        props: () => ({
            // eslint-disable-next-line react-hooks/rules-of-hooks
            controls: useForm(),
            field: {},
        }),
    }));
});
