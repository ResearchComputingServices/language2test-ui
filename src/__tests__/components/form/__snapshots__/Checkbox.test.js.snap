// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Checkbox should match snapshot 1`] = `
<div
  className="field"
>
  <Controller
    as={
      <ForwardRef(WithStyles)
        control={
          <ForwardRef(WithStyles)
            color="primary"
          />
        }
      />
    }
    control={
      Object {
        "defaultValuesRef": Object {
          "current": Object {},
        },
        "dirtyFieldsRef": Object {
          "current": Set {},
        },
        "errorsRef": Object {
          "current": Object {},
        },
        "fieldArrayDefaultValues": Object {
          "current": Object {},
        },
        "fieldArrayNamesRef": Object {
          "current": Set {},
        },
        "fieldsRef": Object {
          "current": Object {},
        },
        "fieldsWithValidationRef": Object {
          "current": Set {},
        },
        "formState": Object {
          "dirty": false,
          "dirtyFields": Set {},
          "isSubmitted": false,
          "isSubmitting": false,
          "isValid": false,
          "submitCount": 0,
          "touched": Object {},
        },
        "getValues": [Function],
        "isDirtyRef": Object {
          "current": false,
        },
        "isWatchAllRef": Object {
          "current": false,
        },
        "mode": Object {
          "isOnBlur": false,
          "isOnChange": false,
          "isOnSubmit": true,
        },
        "reRender": [Function],
        "reValidateMode": Object {
          "isReValidateOnBlur": false,
          "isReValidateOnSubmit": false,
        },
        "readFormStateRef": Object {
          "current": Object {
            "constructor": true,
            "dirty": true,
            "dirtyFields": true,
            "isSubmitted": true,
            "isSubmitting": true,
            "isValid": true,
            "submitCount": true,
            "touched": true,
          },
        },
        "register": [Function],
        "removeFieldEventListener": [Function],
        "resetFieldArrayFunctionRef": Object {
          "current": Object {},
        },
        "setValue": [Function],
        "touchedFieldsRef": Object {
          "current": Object {},
        },
        "triggerValidation": [Function],
        "unregister": [Function],
        "validFieldsRef": Object {
          "current": Set {},
        },
        "watchFieldsRef": Object {
          "current": Set {},
        },
      }
    }
    defaultValue={false}
    rules={
      Object {
        "required": undefined,
      }
    }
  />
</div>
`;
